Improve test coverage for `ServiceIOGroupTest` and `BatchServerTest`:

Krebrov001 opened the pull request:
Improve test coverage for `ServiceIOGroupTest`:
 - `ServiceIOGroupTest.read_signal_exception`
 - `ServiceIOGroupTest.write_control_exception`
 - `ServiceIOGroupTest.save_control`
 - `ServiceIOGroupTest.restore_control`

   - Relates to #2016  [bug report/feature request/story] from github issues

Krebrov001 committed:
Improve test coverage for `ServiceIOGroupTest`:

 - `ServiceIOGroupTest.read_signal_exception`
 - `ServiceIOGroupTest.write_control_exception`

Remove the duplicate test fixtures of `ServiceIOGroupTest`

 - `ServiceIOGroupTest.save_control`
 - `ServiceIOGroupTest.restore_control`

The point of these tests is to show that the method does not do anything.
If we send it an invalid path, we know that if it tries to "do something" i.e. open that file, the test will fail.
The `"/bad/file/path"` is prefered since we can also assume that it cannot be opened.

Also remove some test fixtures that we are not having any more from `test/Makefile.mk`

 - `ServiceIOGroupTest.sample`
 - `ServiceIOGroupTest.adjust`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Adding header comments to test fixtures in `BatchServerTest.cpp`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add `BatchServerTest.stop_batch_exception` test

 - This test checks exceptional circumstances in
`BatchServer::stop_batch()`
   - if it is called in the client process
   - if `sigqueue(2)` fails
 - Adding the `BatchServerTest::m_batch_server_child_process`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Increase coverage by tests that throw exceptions from mock classes.

This is simulating failure of internal components to activate the catch
blocks.

Also added custom `ACTION()` for the `EXPECT_CALL()`

 - `BatchStatusTest.receive_message_exception`
 - `BatchStatusTest.write_message_exception`
 - `BatchStatusTest.read_batch_exception`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add `BatchServerTest.destructor_exceptions`

 - Enable coverage of catch statements in `BatchServerImp::~BatchServerImp()`
   - Test if `BatchServerImp::stop_batch()` can throw an **Exception** in
the destructor upon error conditions.
 - Create the class `cerr_redirect` to facilitate redirecting the stderr to an internal buffer for recording purposes.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Adding extra tests for coverage of case statements in `BatchServerImp::event_loop()`

 - `BatchServerTest.receive_message_terminate`
   - This changes `BatchServerImp::event_loop()` to enable `BatchStatus::M_MESSAGE_TERMINATE` to come over the pipe and stop the `while` loop.
     - This is not the intended use case by the user; this is only for testing purposes.
   - Previously the terminate message was only sent via `SIGTERM` signal, which triggered the signal handler.
 - `BatchServerTest.receive_message_default`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Adding `BatchServerTest.fork_and_terminate_child` and `BatchServerTest.fork_and_terminate_parent`

 - `BatchServerTest::fork_other()` to support switching the server and client between the child process and parent process at will.

These are more than just unit tests in the sense that a fork happens, and there is a child process and a parent process.
One of these processes becomes the server, the other one the client.
These tests check the interaction between the server and the client, when the client sends the server a `BatchStatus::M_MESSAGE_TERMINATE` or `SIGTERM`. So they are more accurate use case than `BatchServerTest.receive_message_terminate`

 - `BatchServerTest.fork_and_terminate_child` checks the situation when the server is on the child process.
   - Simulates the usual way that the `BatchServer` gets a terminate message from the user.
   - NOTE: This test only improves testing of use case. It doesn't improve the coverage. This is because LCOV is not tracking the child process.
 - `BatchServerTest.fork_and_terminate_parent` checks the situation when the server is on the parent process.
   - This has the added benefit that LCOV is able to provide coverage for the lines in the server, such as registering the signal handler, and calling it when a `SIGTERM` is received from the client.

Here we are also using a custom mock for `m_batch_status->receive_message()` takes a lambda function as a behavior which simulates blocking on the `read()` system call, and then throws an `geopm::Exception` when the `SIGTERM` is received.

Misc edits to `BatchServerTest.cpp`

 - add `EXPECT_CALL()` for `m_batch_status->send_message()` in `BatchServerTest.read_batch_exception`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Adding `BatchServerTest.fork_with_setup_exception`

 - This test provides coverage for the error case in `BatchServerImp::fork_with_setup()`, when something over than `BatchStatus::M_MESSAGE_CONTINUE` is sent over the pipe, it would cause an `Exception` to be thrown.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +788,466 @@ TEST_F(BatchServerTest, fork_with_setup)
+                /* Extra code for synchronizing the server. */
+               char unique_char = '!';
+               int ret = write(write_pipe_fd, &unique_char, sizeof(unique_char));
+               ret = ret;
```
```suggestion
```

This line is breaking the clang build.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +788,466 @@ TEST_F(BatchServerTest, fork_with_setup)
+
+    /// This function contains the parent process, which is the client.
+    /// It takes the read_pipe_fd and the PID of the child process, which is the server.
+    std::function<void(int, int)> parent_process_func = [this](int read_pipe_fd, int server_pid)
```
```suggestion
    std::function<void(int, int)> parent_process_func = [](int read_pipe_fd, int server_pid)
```

Unless you're going to use `this`, it must be omitted since warnings are treated as errors.  It's an unused variable otherwise.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +788,466 @@ TEST_F(BatchServerTest, fork_with_setup)
+                /* Extra code for synchronizing the server. */
+               char unique_char = '!';
+               int ret = write(write_pipe_fd, &unique_char, sizeof(unique_char));
+               ret = ret;
```
```suggestion
```

This is also breaking the clang build.

bgeltz reviewed (changes requested):
Please address these compilation issues.

Krebrov001 committed:
Resolve leaked mock objects using `_Exit()` instead of `testing::Mock::AllowLeak()` and `this->TearDown()`

The underlying problem was that **Google Test** was reporting leaked mock objects in the forked process (child process). To mitigate this we previously used `testing::Mock::AllowLeak()` and `this->TearDown()`. Now instead we are using `_Exit()` in the forked process, which stops reporting leaked mock objects.
The `_Exit()` function in C/C++ gives normal termination of a program without performing any cleanup tasks.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on service/test/ServiceIOGroupTest.cpp:
```diff
@@ -181,6 +181,37 @@ TEST_F(ServiceIOGroupTest, read_signal_behavior)
+{
+    // !is_valid_signal(signal_name)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_serviceio_group->read_signal("NUM_VACCUM_TUBES", 4, 0),
```
```suggestion
        m_serviceio_group->read_signal("NUM_VACUUM_TUBES", 4, 0),
```

I know it's a nit, but it's spelled "vacuum".  This does not affect the functionality.

bgeltz commented on service/test/ServiceIOGroupTest.cpp:
```diff
@@ -192,6 +223,39 @@ TEST_F(ServiceIOGroupTest, write_control)
+{
+    // !is_valid_control(control_name)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_serviceio_group->write_control("NUM_VACCUM_TUBES", 4, 0, 7.0),
```
```suggestion
        m_serviceio_group->write_control("NUM_VACUUM_TUBES", 4, 0, 7.0),
```

I know it's a nit, but it's spelled "vacuum".  This does not affect the functionality.

bgeltz commented on service/test/ServiceIOGroupTest.cpp:
```diff
@@ -264,13 +328,15 @@ TEST_F(ServiceIOGroupTest, write_batch)
 TEST_F(ServiceIOGroupTest, save_control)
 {
     /// These are noops, make sure mock objects are not called into
+    /// We want it to fail if it tries to open the non-existent file path.
```
Where are we testing that this failure actually happens?  I see we open a bad path here, but I also expected something like ```GEOPM_EXPECT_THROW_MESSAGE(...)```.

Same comment for the `restore_control` test below.

cmcantalupo replied:
Note we are checking that the ServiceIOGroup::save_control() method does not do anything.  We provide the bad file path to make sure that the method does not try to use the path.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -47,6 +48,11 @@
+#include <cstring>    // for strlen(), strncmp()
+#include <string>     // for std::string
+#include <sstream>    // for std::stringstream
+#include <iostream>   // for std::streambuf, std::cerr, std::flush
```
```suggestion
#include <cstring>
#include <string>
#include <sstream>
#include <iostream>
```

These comments are unnecessary.  Please remove them.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -63,6 +69,7 @@ class BatchServerTest : public ::testing::Test
     protected:
         void SetUp();
         void TearDown();
+        int fork_other(std::function<void(int, int)> child_process_func, std::function<void(int, int)> parent_process_func, bool child_is_server);
```
```suggestion
        int fork_other(std::function<void(int, int)> child_process_func,
                       std::function<void(int, int)> parent_process_func,
                       bool child_is_server);
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -150,8 +168,12 @@ void BatchServerTest::TearDown()
 }
 
+/**
+ * @test BatchServerImp::server_pid() in isolation
```
```suggestion
 * @test Check BatchServerImp::server_pid() in isolation
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -160,12 +182,18 @@ TEST_F(BatchServerTest, get_server_pid)
 }
 
+/**
+ * @test BatchServerImp::get_server_key() in isolation
```
```suggestion
 * @test Check BatchServerImp::get_server_key() in isolation
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -160,12 +182,18 @@ TEST_F(BatchServerTest, get_server_pid)
 }
 
+/**
+ * @test Check BatchServerImp::stop_batch() to send SIGTERM signal to the server.
```
```suggestion
 * @test Check that BatchServerImp::stop_batch() sends a SIGTERM signal to the server.
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -177,6 +205,57 @@ TEST_F(BatchServerTest, stop_batch)
 }
 
+/**
+ * @throws geopm Exception to simulate failure of sigqueue(2) system call in POSIXSignalImp::sig_queue()
```
```suggestion
 * @throws geopm::Exception() to simulate a failure of the sigqueue(2) system call in POSIXSignalImp::sig_queue()
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -177,6 +205,57 @@ TEST_F(BatchServerTest, stop_batch)
+
+/**
+ * @test First check that BatchServerImp::stop_batch() throws an exception if it is called from child process.
+ *       Then check that activate the catch block in BatchServerImp::stop_batch() by failure of POSIXSignalImp::sig_queue()
```
```suggestion
 * @test Check that BatchServerImp::stop_batch() throws an exception if it is called from the child process.
 *       Then check the catch block in BatchServerImp::stop_batch() by forcing a failure of POSIXSignalImp::sig_queue().
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -235,6 +313,14 @@ TEST_F(BatchServerTest, run_batch_read)
 
+/**
+ * @test Check BatchServerImp::run_batch() when there are no signals and you try to read.
+ *       First the requests only controls are filled up.
```
```suggestion
 *       First the control requests are populated.
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
```
```suggestion
```
Not used.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+
+/**
+ * @test Provides coverage for the case when not a valid message is read
+ *       by the server in BatchServerImp::event_loop()
```
```suggestion
 * @test Provides coverage for the case when an invalid message is read
 *       by the server in the BatchServerImp::event_loop().
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        m_batch_server->run_batch(),
+        GEOPM_ERROR_RUNTIME,
+        "BatchServerImp::run_batch(): Received unknown response from client: "
+    );
```
If the unknown response message (127) is in the exception text, then we should verify it is there.

It would be good if the "random" status message (127) was stored in a variable and that variable was used with `GEOPM_EXPECT_THROW_MESSAGE` to ensure that the proper unknown response was received (127 in this case).

The way the test is now it asserts that *some* exception is generated because of an unknown response, but we are specifically sending 127, so we should verify that.

If the 127 does not already show up in the exception text, then ignore this comment.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+}
+
+/**
+ * @throws geopm Exception to simulate failure of read(2) system call in BatchStatusImp::receive_message()
```
```suggestion
 * @throws geopm::Exception() to simulate a failure of the read(2) system call in BatchStatusImp::receive_message().
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
```
```suggestion
```
Not used.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
```
```suggestion
```
Not used.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+    InSequence sequence;
+
+    int idx = 0;
+    std::vector<double> result = {240.042, 250.052};
```
```suggestion
```
Not used.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+        ++idx;
+    }
+
+    for (const auto &request : m_control_config) {
```
```suggestion
    idx = 0;
    for (const auto &request : m_control_config) {
```

When pushing signals and controls, make sure the index is reset in between.  It doesn't matter for this test because you are not verifying the indices, but it will lead to confusing behavior if this code is used in other tests that do utilizes the indices.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +461,263 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+
+/**
+ * @test Check if we can create the shared memory.
+ *       Check if the shared memory files for both the signals and controls were created in the /dev/shm/
```
```suggestion
 *       Check if the shared memory files for both the signals and controls were created in the /dev/shm/.
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -177,6 +205,57 @@ TEST_F(BatchServerTest, stop_batch)
+        EINVAL,
+        __FILE__,
+        __LINE__
+    );
```
```suggestion
    throw geopm::Exception("POSIXSignal(): POSIX signal function call sigqueue() returned an error",
                           EINVAL, __FILE__, __LINE__);
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -177,6 +205,57 @@ TEST_F(BatchServerTest, stop_batch)
+ */
+TEST_F(BatchServerTest, stop_batch_exception)
+{
+    GEOPM_EXPECT_THROW_MESSAGE(
+        m_batch_server_child_process->stop_batch(),
+        GEOPM_ERROR_RUNTIME,
+        "BatchServerImp::stop_batch(): must be called from parent process, not child process"
+    );
```
```suggestion
    GEOPM_EXPECT_THROW_MESSAGE(m_batch_server_child_process->stop_batch(),
                               GEOPM_ERROR_RUNTIME, 
                               "BatchServerImp::stop_batch(): must be called from parent process, not child process");
```

bgeltz replied:
Please use this style for spacing and line termination for function calls.  Please let me know if you have questions.

There are many calls to `GEOPM_EXPECT_THROW_MESSAGE` in this file.  Please fix all of them.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+    {
+        // return something other than BatchStatus::M_MESSAGE_CONTINUE
+        // so that it would write that bad message to the pipe in the child process
+        // and the parent process would throw the Exception.
```
```suggestion
```

That comment is redundant. The information is already discussed in the docstring.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+    GEOPM_EXPECT_THROW_MESSAGE(
+        (forked_pid = m_batch_server_pid->fork_with_setup(setup, run)),
+        GEOPM_ERROR_RUNTIME,
+        "BatchServerImp: Receivied unexpected message from batch server at startup: \""
```
Can you add to the expected text here about the specific unexpected message (M_MESSAGE_QUIT) that was received?  Is that text already provided in the exception similar to my other comment?

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+ *
+ * @details This class facilitates redirect the stderr to an internal buffer for recording purposes.
+ */
+class cerr_redirect {
```
```suggestion
class CerrRedirect {
```
Use camel case for C++ classes.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        /**
+         * @brief The redirect is NOT active when the object is created.
+         */
+        cerr_redirect()
```
```suggestion
        CerrRedirect()
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        /**
+         * @brief Restore the backup streambuf if it was not restored already.
+         */
+        ~cerr_redirect()
```
```suggestion
        ~CerrRedirect()
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+                buffer.str(std::string());
+                // Set the streambuf of the std::cerr equivalent to the streambuf of the stringstream.
+                // Return the former streambuf of the std::cerr prior to the change.
+                cerr_backup = std::cerr.rdbuf( buffer.rdbuf() );
```
```suggestion
                cerr_backup = std::cerr.rdbuf(buffer.rdbuf());
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+                is_active = false;
+
+                return buffer.str();
+            } else {
```
```suggestion
            } 
            else {
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        std::string close_redirect(void)
+        {
+            if (is_active)
+            {
```
```suggestion
            if (is_active) {
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+            }
+        }
+    private:
+        /// records the redirected stderr into it's own streambuf
```
You used the triple slash here to indicate that it's a docstring, but then there's no @keyword here.  Please insert the missing @keyword or remove these lines.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+ACTION(sig_queue_throw_2)
+{
+    throw 2;
+}
```
It seems odd to me to mix these the usage of these ACTION's with the usage of lambda's.  Both are used in this file, and I think it would be good to use one or the other.  My preference is with lambda's.  I think this is one we should discuss as a group with @cmcantalupo.

The usage of these ACTIONs doesn't add anything new AFAICT.  It's just more complexity.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+    ///
+
+    // Create new BatchServer object
+    std::shared_ptr<BatchServerImp> m_batch_server_exception = std::make_shared<BatchServerImp>(
```
```suggestion
    std::shared_ptr<BatchServerImp> batch_server_exception = std::make_shared<BatchServerImp>(
```
This variable is not a class member.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        m_signal_shmem,
+        m_control_shmem,
+        m_server_pid
+    );
```
```suggestion
        m_server_pid);
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        sig_queue(m_server_pid, SIGTERM, BatchStatus::M_MESSAGE_TERMINATE)
+    )
+    .Times(1)
+    .WillRepeatedly(sig_queue_SIGTERM());
```
```suggestion
    EXPECT_CALL(*m_posix_signal.get(),
                sig_queue(m_server_pid, SIGTERM, BatchStatus::M_MESSAGE_TERMINATE))
        .Times(1)
        .WillRepeatedly(sig_queue_SIGTERM());
```
Please follow the style of other invocations of EXPECT_CALL.  They should all look similar to this.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+    m_batch_server_exception.reset();  // calling BatchServerImp::~BatchServerImp()
+    std::string actual_stderr_1 = c_redirect.close_redirect();
+    // Compare if the recorded stderr matches the expected stderr.
+    EXPECT_EQ(0, strncmp(expected_stderr_1, actual_stderr_1.c_str(), stderr_1_length));
```
Since you're using `std::string`'s in the CerrRedirect class, it seems logical to stick with them and not use a mix of `std::string`'s and C strings.

I would change all of this code to use `std::string`'s and then remove all of the C comparison functions and use `EXPECT_EQ` with `std::string` objects.  This is the pattern we use in other tests.
https://google.github.io/googletest/reference/assertions.html#c-strings

This will also cleanup your `#include`'s.  Our general rule is to avoid using C-style code unless it's absolutely necessary for interacting with some API.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        m_signal_shmem,
+        m_control_shmem,
+        m_server_pid
+    ));
```
```suggestion
        m_server_pid));
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        sig_queue(m_server_pid, SIGTERM, BatchStatus::M_MESSAGE_TERMINATE)
+    )
+    .Times(1)
+    .WillRepeatedly(sig_queue_throw_2());
```
Wrong style.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        }
+        // Use _Exit() instead of exit() to prevent Google Test from complaining about
+        // leaked mock objects.
+        _Exit(EXIT_SUCCESS);
```
I want to understand why this is necessary.

cmcantalupo replied:
_Exit() should be called by a forked process to allow the statically allocated variables to be destroyed by the parent process, not by the forked process.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        // leaked mock objects.
+        _Exit(EXIT_SUCCESS);
+    // parent process //
+    } else {
```
```suggestion
    }
    else {
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+            close(write_pipe_fd);  // close write end of pipe
+            parent_process_func(read_pipe_fd, result);  // pass read end of pipe
+            close(read_pipe_fd);  // close read end of pipe
+        } else {  // parent is server
```
```suggestion
        }
        else {  // parent is server
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+            nullptr,           // Create a real POSIXSignal because we want to use sig_action()
+            m_signal_shmem,
+            m_control_shmem,
+            m_server_pid
+        );
```
```suggestion
            m_server_pid);
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        m_batch_server_test->child_register_handler();
+
+        int idx = 0;
+
```
```suggestion
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +790,456 @@ TEST_F(BatchServerTest, fork_with_setup)
+        ++idx;
+        }
+
+        for (const auto &request : m_control_config) {
```
```suggestion
        idx = 0;
        for (const auto &request : m_control_config) {
```

Krebrov001 committed:
Addressing reviewers feedback

 - Fixing mostly formatting and style.
 - Replace the usage of ACTION's with the usage of lambda's to specify the behaviors of methods of the mocked objects.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on service/test/ServiceIOGroupTest.cpp:
```diff
@@ -181,6 +181,37 @@ TEST_F(ServiceIOGroupTest, read_signal_behavior)
+        m_serviceio_group->read_signal("NUM_VACUUM_TUBES", 4, 0),
+        GEOPM_ERROR_INVALID,
+        "ServiceIOGroup::read_signal(): signal name \""
+    );
```
```suggestion
    GEOPM_EXPECT_THROW_MESSAGE(m_serviceio_group->read_signal("NUM_VACUUM_TUBES", 4, 0),
                               GEOPM_ERROR_INVALID, "ServiceIOGroup::read_signal(): signal name");
```

Please use this style for spacing and line endings.  Please change all invocations of ```GEOPM_EXPECT_THROW_MESSAGE``` to match this style.  Look at how other tests use ```GEOPM_EXPECT_THROW_MESSAGE``` and let me know if you have questions about the style.

Additionally, I changed the string in the last argument slightly.  I removed '\"'.  It seems like you started to set this up to expect a specific message in the exception text (I'm assuming "NUM_VACUUM_TUBES"), but then stopped.  What was the intention here?

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -160,39 +179,77 @@ TEST_F(BatchServerTest, get_server_pid)
+                throw geopm::Exception("POSIXSignal(): POSIX signal function call sigqueue() returned an error",
+                           EINVAL, __FILE__, __LINE__);
+            }
+        );
```
```suggestion
        .WillRepeatedly(Throw(geopm::Exception("POSIXSignal(): POSIX signal function call sigqueue() returned an error",
                              EINVAL, __FILE__, __LINE__)));
```

Please remove all these lambdas and use the ```Throw()``` action.
https://google.github.io/googletest/reference/actions.html

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +437,227 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+
+    GEOPM_EXPECT_THROW_MESSAGE(m_batch_server->run_batch(),
+                               GEOPM_ERROR_RUNTIME,
+                               "BatchServerImp::run_batch(): Received unknown response from client: " "127");
```
Please move the random number (127) into a variable, and use that variable with ```EXPECT_CALL``` and ```GEOPM_EXPECT_THROW_MESSAGE```.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +437,227 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+                throw geopm::Exception("BatchStatusImp: System call failed: " "read(2)",
+                           EINTR, __FILE__, __LINE__);
+                return '\0';
+            }
```
Remove this lambda and use the Throw action similar to the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +437,227 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+                throw geopm::Exception("BatchStatusImp: System call failed: " "write(2)",
+                           EINTR, __FILE__, __LINE__);
+            }
+        );
```
Remove this lambda and use the Throw action similar to the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +437,227 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+
+    GEOPM_EXPECT_THROW_MESSAGE(m_batch_server->run_batch(),
+                               EINTR,
+                               "BatchStatusImp: System call failed: " "write(2)");
```
```suggestion
                               "BatchStatusImp: System call failed: write(2)");
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -359,6 +437,227 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+            [](){
+                throw geopm::Exception("SSTIOImp::read_batch() mbox read failed",
+                           EINVAL, __FILE__, __LINE__);
+            }
```
Remove this lambda and use the Throw action similar to the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+    int forked_pid = 0;
+    GEOPM_EXPECT_THROW_MESSAGE((forked_pid = m_batch_server_pid->fork_with_setup(setup, run)),
+                               GEOPM_ERROR_RUNTIME,
+                               "BatchServerImp: Receivied unexpected message from batch server at startup: \"" "113" "\"");
```
```suggestion
                               "BatchServerImp: Receivied unexpected message from batch server at startup: \" 113\"");
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+        // the backup streambuf associated with the stderr
+        std::streambuf* cerr_backup;
+        // is true if the redirect is active
+        bool is_active;
```
All class members should start with "m_":
```suggestion
        bool m_is_active;
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+        // records the redirected stderr into it's own streambuf
+        std::stringstream buffer;
+        // the backup streambuf associated with the stderr
+        std::streambuf* cerr_backup;
```
```suggestion
        std::streambuf* m_cerr_backup;
```
All class members should start with "m_":

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+        }
+    private:
+        // records the redirected stderr into it's own streambuf
+        std::stringstream buffer;
```
```suggestion
        std::stringstream m_buffer;
```
All class members should start with "m_":

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+        m_posix_signal,
+        m_signal_shmem,
+        m_control_shmem,
+        m_server_pid);
```
Please follow this style when there are a large number of arguments:
https://github.com/geopm/geopm/blob/07334f10777d04914179bb8d8649f3b6d6f92488/service/test/BatchServerTest.cpp#L94-L102

Please change all occurrences if there are multiple.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+    batch_server_exception.reset();  // calling BatchServerImp::~BatchServerImp()
+    std::string actual_stderr_1 = c_redirect.close_redirect();
+    // Compare if the recorded stderr matches the expected stderr.
+    EXPECT_EQ(true, geopm::string_begins_with(actual_stderr_1, expected_stderr_1));
```
Use ```EXPECT_TRUE``` for assertions like this.
https://google.github.io/googletest/reference/assertions.html

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+    batch_server_exception.reset();  // calling BatchServerImp::~BatchServerImp()
+    actual_stderr_1 = c_redirect.close_redirect();
+    // Compare if the recorded stderr matches the expected stderr.
+    EXPECT_STREQ(expected_stderr_1.c_str(), actual_stderr_1.c_str());
```
There's no need to use C strings nor ```EXPECT_STREQ```.  

Instead, use ```EXPECT_EQ``` and the string object.
https://google.github.io/googletest/reference/assertions.html#c-strings

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+            [](){
+                throw geopm::Exception("POSIXSignal(): POSIX signal function call sigqueue() returned an error",
+                           EINVAL, __FILE__, __LINE__);
+            }
```
Remove this lambda and use the Throw action similar to the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+            /// @throws (int)2 to simulate failure of sigqueue(2) system call in POSIXSignalImp::sig_queue()
+            [](){
+                throw 2;
+            }
```
Remove this lambda and use the Throw action similar to the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
```
```suggestion
            ++idx;
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
```
```suggestion
            ++idx;
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+                return '$';
+            }
+        )
+        .RetiresOnSaturation();
```
```suggestion
        EXPECT_CALL(*m_batch_status, receive_message())
            .WillOnce([&write_pipe_fd]() {
                /* Extra code for synchronizing the server. */
                char unique_char = '!';
                write(write_pipe_fd, &unique_char, sizeof(unique_char));
                sleep(1024);
                throw geopm::Exception("BatchStatusImp: System call failed: " "read(2)",
                                       EINTR, __FILE__, __LINE__);
                // This is irrelevant, it doesn't get returned because of the throw statement.
                // This is just to match the return type of BatchStatusImp::receive_message()
                return '$';
            }).RetiresOnSaturation();
```

For any lambdas that cannot be converted to ```Throw``` actions, please follow this style.  Here's an example from another test:
https://github.com/geopm/geopm/blob/07334f10777d04914179bb8d8649f3b6d6f92488/service/test/SSTIOTest.cpp#L229-L233

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+                                            request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
```
```suggestion
            ++idx;
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+                                             request.domain_idx))
+            .WillOnce(Return(idx))
+            .RetiresOnSaturation();
+        ++idx;
```
```suggestion
            ++idx;
```

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+        // and receive_message() throws an exception in the check_return().
+        // This exception is caught in the BatchServerImp::read_message()
+        // and char in_message = BatchStatus::M_MESSAGE_TERMINATE is returned to BatchServerImp::event_loop()
+        EXPECT_CALL(*m_batch_status, receive_message())
+        .WillOnce(
+            [&write_pipe_fd](){
+                /* Extra code for synchronizing the server. */
+               char unique_char = '!';
+               write(write_pipe_fd, &unique_char, sizeof(unique_char));
+
+                sleep(1024);
+                throw geopm::Exception("BatchStatusImp: System call failed: " "read(2)",
+                                       EINTR, __FILE__, __LINE__);
+
+                // This is irrelevant, it doesn't get returned because of the throw statement.
+                // This is just to match the return type of BatchStatusImp::receive_message()
+                return '$';
+            }
+        )
+        .RetiresOnSaturation();
```
Please change this to follow the style to the one above.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -426,3 +730,441 @@ TEST_F(BatchServerTest, fork_with_setup)
+    std::function<void(int, int)> parent_process_func = [this](int write_pipe_fd, int client_pid)
+    {
+        // Create new BatchServer object
+        std::shared_ptr<BatchServerImp> batch_server_test = std::make_shared<BatchServerImp>(
+            client_pid,
+            m_signal_config,
+            m_control_config,
+            *m_pio_ptr,
+            m_batch_status,
+            nullptr,           // Create a real POSIXSignal because we want to use sig_action()
+            m_signal_shmem,
+            m_control_shmem,
+            m_server_pid);
```
Please update this to follow the style of the above suggestion.

bgeltz commented on service/test/BatchServerTest.cpp:
```diff
@@ -271,21 +331,29 @@ TEST_F(BatchServerTest, run_batch_read_empty)
     }
 
-     for (const auto &request : m_control_config) {
+    // idx should be the same when we use write_batch()
```
```suggestion
```

This comment doesn't make sense to me.  What is requiring this to be true?

bgeltz reviewed (changes requested):
This is starting to look pretty good now.

I'm good to merge after either dropping or justifying that comment.

Krebrov001 committed:
Add new tests for coverage of `BatchServer` `SIGCHLD` handler.

 - `BatchServerTest.action_sigchld` for normal operation
 - `BatchServerTest.action_sigchld_error` for error condition

Also added `geopm::error_message()` for easily converting geopm error
messages into their corresponding string representation.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -69,6 +69,9 @@ static void action_sigchld(int signo, siginfo_t *siginfo, void *context)
             g_sigchld_status = child_status;
         }
     }
+    if (child_pid == -1) {
+        g_sigchld_status = GEOPM_ERROR_RUNTIME;
+    }
```
```suggestion
    if (child_pid == -1) {
        g_sigchld_status = errno ? errno : GEOPM_ERROR_RUNTIME;
    }
```

cmcantalupo commented on service/src/geopm/Exception.hpp:
```diff
@@ -119,6 +119,11 @@ namespace geopm
             int m_err;
     };
+
+    /// @param error_value The error code associated with the exception.
```
```suggestion
    /// @brief Function that converts an error code into an error message
    /// @param error_value The error code associated with the exception.
```

cmcantalupo replied:
There is an equivalent C interface here:

https://github.com/geopm/geopm/blob/dev/service/src/geopm_error.h#L60

perhaps it would be best to match the name:

geopm::get_error_string() -> geopm::error_message()

bgeltz replied:
The interface Chris is referencing is implemented here:
https://github.com/geopm/geopm/blob/dev/service/src/Exception.cpp#L71

@cmcantalupo Why have two different implementations though?  My thought was to change the test to use the C interface and remove this one.

cmcantalupo replied:
It is useful to be able to directly use the std::string type and not have a fixed length C string.  Additionally the C interface provides the message from the last handled exception (useful for our external C interfaces), but this is not helpful in a context where there may not be a handled exception.

cmcantalupo reviewed (commented):
Couple of minor suggestions.

cmcantalupo reviewed (dismissed):
Thanks! Looks good.  Please adjust the history as you see fit, and then we can merge it in.

Krebrov001 commented:
Pull request successfully merged into `dev` branch.
